Add fan and switch support

webdjoe opened the pull request.

webdjoe committed:
Add fan and switch support

markperdue commented on src/pyvesync/vesyncoutlet.py:
```diff
@@ -19,52 +19,76 @@ def __init__(self, details, manager):
     def active_time(self):
         """Return active time of a device in minutes"""
+        if 'active_time' not in self.details:
+            self.get_details()
```
I learned a lesson on doing what you are trying to do here awhile ago. The HA team doesn't like it when anything that is expected to simply read and return a value makes a call out to an API to get the data if it doesn't exist. The library used to do exactly what you're doing here and since the HA integration was a passthrough to get the energy_today, power, and these types of values a hasty PR was made against the HA vesync integration that bypassed the library function that was causing updates

webdjoe replied:
Your right, I remember reading the convo in your original PR. I'll fix it

markperdue commented on src/pyvesync/vesyncoutlet.py:
```diff
@@ -82,8 +106,12 @@ def get_details(self):
-            self.details['power'] = r.get('power')
-            self.details['voltage'] = r.get('voltage')
+            power = r.get('power', '0:0')
+            power = round(float(helpers.calculate_hex(power)), 2)
```
nice catch

markperdue commented on src/pyvesync/vesyncoutlet.py:
```diff
@@ -163,26 +191,16 @@ def turn_off(self):
     def get_body(self, type_):
         if type_ == 'detail':
-            body = helpers.req_body('devicedetail')
+            body = helpers.req_body(self.manager, 'devicedetail')
```
nice catches

markperdue commented on src/pyvesync/vesyncbasedevice.py:
```diff
@@ -1,3 +1,6 @@
+from abc import ABC, ABCMeta
```
unneeded import at this time

webdjoe replied:
i originally had VeSyncBaseDevice as an ABCMeta but there aren't any abstractmethods so the only reason to have it would be to prevent it from being instantiated.  I wanted to get your opinion on this and test as is first.

markperdue commented on src/pyvesync/vesyncbasedevice.py:
```diff
@@ -42,3 +45,10 @@ def configure(self, details):
         self.extension = details.get('extension', None)
         self.current_firm_version = details.get('currentFirmVersion', None)
+
+        if self.device_type is not None:
```
what does this do? future functionality?

webdjoe replied:
This is the part of a test I made to ensure the device is configured correctly but I figured it would be redundant if the devtype in the factory already determines the device class. Consider it gone

webdjoe committed:
Update interval, minor corrections, cleanup

Add support to define update interval in VeSync class and checking for an interval when running outlet.update_energy()

markperdue reviewed (approved):
changes look good. nice work
